### PR TITLE
XML Mapping for Hibernate ORM

### LINKED ISSUE #14762: JPA orm.xml file handling to support multiple database vendors
**Description**
I need to implement a solution that supports multiple database vendors like Oracle, MS SQLServer, DB2, etc.

Moreover I need to re-use existing "legacy" JARs containing JPA entities with annotations and orm.xml files.
Such "legacy" JARs are used by existing Java EE 8 applications.

Existing database schemas use sequences for Oracle PK and identities for SQLServer PK.
To handle such differences, we use annotations to define Oracle sequences .. and we use the orm.xml file to define MS SQLServer identities.

Moreover the orm.xml file is used for setting persistence-unit/schema as well. It means it is not only a PK override issue.

The orm.xml file is a standard JPA feature that is very useful (imho) and I believe this file support is mandatory for such use-cases.

Please keep in mind that migration of legacy apps/libraries is an important point here.

Thank you!

### PR DESCRIPTION
**Pull request by yrodiere**
Fixes #14762

Creating as a draft, because we need to upgrade to Hibernate ORM 5.5 first (#17010).

### CODE REVIEW

**Comment by yrodiere** [comment]
Ready for review now that the upgrade to ORM 5.5 was merged.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmConfigPersistenceUnit.java
Comment (machi1990) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmConfigPersistenceUnit.java:122]: Do we hot reload these files?

Comment (yrodiere) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmConfigPersistenceUnit.java:122]: I have absolutely no idea. I didn't do anything special to have them hot-reloaded, anyway.

They are processed in a build step: `io.quarkus.hibernate.orm.deployment.HibernateOrmProcessor#defineJpaEntities`.

I will add a test regarding hot-reload of XML mapping files. If it doesn't work, any hint on how I can tell Quarkus that some files are relevant to hot-reload?

Comment (machi1990) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmConfigPersistenceUnit.java:122]: Looking at the PR, they are not enabled for hot reload. To do so, you might want to produce a `HotDeploymentWatchedFileBuildItem` just like https://github.com/quarkusio/quarkus/blob/ac05c00115efc9008e743ce6036a6d435687f5de/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java#L200 or https://github.com/quarkusio/quarkus/blob/ac05c00115efc9008e743ce6036a6d435687f5de/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java#L931

Comment (yrodiere) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmConfigPersistenceUnit.java:122]: Done! Thanks for the tip.

**Comment by yrodiere** [comment]
I pushed an update to enable hot-reload for XML mapping files.

Also I tried to fix tests failing on CI, apparently because of a wrong username when connection to the database (?!). I cannot reproduce the error locally, though, so fingers crossed...

**Comment by Sanne** [comment]
> Also I tried to fix tests failing on CI, apparently because of a wrong username when connection to the database (?!). I cannot reproduce the error locally, though, so fingers crossed...

which DB was that?

**Comment by yrodiere** [comment]
> > Also I tried to fix tests failing on CI, apparently because of a wrong username when connection to the database (?!). I cannot reproduce the error locally, though, so fingers crossed...
> 
> which DB was that?

H2, can you believe it? :)

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmConfigPersistenceUnit.java
Comment (machi1990) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmConfigPersistenceUnit.java:122]: Awesome, thanks. You are welcome.

**Comment by yrodiere** [comment]
> Also I tried to fix tests failing on CI, apparently because of a wrong username when connecting to the database (?!). I cannot reproduce the error locally, though, so fingers crossed...

Didn't work, unfortunately... I'm on it.

**Comment by Sanne** [comment]
@yrodiere not meaning to steal the fun, but I happened to notice.. the `H2DevServicesProcessor` starts the H2 database expecting password "sa" but the configuration expects an empty password.

Incidentally, not sure why `H2DevServicesProcessor` has "postgresql" in its package name...

**Comment by yrodiere** [comment]
Oh don't worry, this is no fun at all :)

I think I found the culprit: there was a very dump mistake (not renaming the `application-*.properties` to `application.properties` when building the archive to test). This led to using dev-services when running my tests locally, which worked, but for some reason didn't work on CI (who knows why).

I fixed the configuration, let's see how it goes now...

**Comment by Sanne** [comment]
Great, the remaining failing test is certainly unrelated.

I had a good look at the PR today but couldn't finish it - sorry let's continue tomorrow, looks great so far.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/NonJpaModelBuildItem.java
Comment (Sanne) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/NonJpaModelBuildItem.java:10]: @FroMage I do remember you had to introduce this, but I don't remember why... seems like Yoann is right and this is dead code now, seems safe to remove. Are we missing something?

Comment (yrodiere) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/NonJpaModelBuildItem.java:10]: @Sanne It's worse than "dead code now", from what I can see in the git log it has always been dead code.

Comment (FroMage) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/NonJpaModelBuildItem.java:10]: Safe to remove. It was made for when Reactive Panache had its own JPA-lite implementation.

Comment (FroMage) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/NonJpaModelBuildItem.java:10]: It was not dead code, though, it was used in a branch.

Comment (FroMage) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/NonJpaModelBuildItem.java:10]: But that branch is now dead.

**Comment by Sanne** [comment]
Looks all good, but rebased it so to solve conflicts.

**Review by Sanne — APPROVED** [review]

File: core/deployment/src/main/java/io/quarkus/deployment/recording/PropertyUtils.java
Comment (Sanne) [inline core/deployment/src/main/java/io/quarkus/deployment/recording/PropertyUtils.java:36]: This one change makes the whole CI test much slower, as all modules depend on it.

Since it looks like we'll have to run the tests several more times, might be easier of you submit this particular change as a separate PR first?

**Review by Sanne — CHANGES_REQUESTED** [review]
I'm also seeing local issues with metaspace now, and it looks like caused by changes in this PR (it seems to work fine otherwise).

I'm afraid we'll have to get a better look at that before merging this

**Comment by Sanne** [comment]
hi @yrodiere , I was having another look at the metaspace issue - I couldn't figure out yet what exactly is going on, but I narrowed it down to code being generated by test `HibernateET`. (I don't understand this test purpose ?)

Part of the problem seems to be that this test fails to boot complaining about duplicate entity names in the project; I did a quick and dirt refactoring to avoid duplicate names and the metaspace problem seems gone.

Want to try? cherry pick this from my fork: https://github.com/quarkusio/quarkus/commit/917090359fca4fb06fb2c34ec30ee241b86c8dd4

File: core/deployment/src/main/java/io/quarkus/deployment/recording/PropertyUtils.java
Comment (yrodiere) [reply core/deployment/src/main/java/io/quarkus/deployment/recording/PropertyUtils.java:36]: I'll have a look, though we'd need to get it merged fast in order for the separate PR to be useful... And that separate PR will have the exact same problem of needing to rebuild everything :)

Comment (Sanne) [reply core/deployment/src/main/java/io/quarkus/deployment/recording/PropertyUtils.java:36]: Yes but as soon as you send it I will approve it :)

**Comment by Sanne** [comment]
this also relates: #17277

**Comment by yrodiere** [comment]
@Sanne I believe you when you say your patch fixes the problem, but the root cause looks like a bug in Quarkus itself... I wouldn't expect those other entity types to be picked up in this test. Even if it's normal, I'd rather configure this test explicitly to only take *some* entity types into account (the ones this test relies on), rather than savagely include every single entity type for no reason.

I'll investigate a bit more.

File: core/deployment/src/main/java/io/quarkus/deployment/recording/PropertyUtils.java
Comment (yrodiere) [reply core/deployment/src/main/java/io/quarkus/deployment/recording/PropertyUtils.java:36]: https://github.com/quarkusio/quarkus/pull/17279

**Comment by yrodiere** [comment]
> I'll investigate a bit more.

From my understanding `HibernateET` is not supposed to be run "as is" but only from `HibernateHotReloadTestCase`... and in that context I think it only picks up the one entity type defined in  `HibernateHotReloadTestCase`. So I don't think entity name conflicts are related to our problems at all, they are just something that happens when you run `HibernateET` manually (which you shouldn't).

Let's wait for Stuart's opinion on the "general leak" (https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/HibernateET)?

**Comment by yrodiere** [comment]
Rebased on #17279 just now.

**Comment by Sanne** [comment]
> @Sanne I believe you when you say your patch fixes the problem, but the root cause looks like a bug in Quarkus itself... I wouldn't expect those other entity types to be picked up in this test.

Sure, I agree. It all seems to point to an existing issues, which implies working around it is acceptable to make progress with your PR.

>  Even if it's normal, I'd rather configure this test explicitly to only take _some_ entity types into account (the ones this test relies on), rather than savagely including every single entity type for no reason.

It already does, which is why I'm pointing this out: there seems to be a relation.
 
> I'll investigate a bit more.

Thanks that's nice. Obviously ideally I prefer we figure it out, but alternatively you could break it up and entrust that specific investigation to a different issue / person.

**Comment by yrodiere** [comment]
With your patch @Sanne the tests that would previously fail now hang forever... I tried removing the seemingly unrelated changes to the number of HTTP IO threads, and then the metaspace error comes back, unfortunately...

I think we'll need @stuartwdouglas's help with on the "general leak" (https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/HibernateET), which he said he could provide tomorrow (maybe).

Reproducer: simply build this PR, then execute the tests from `extensions/hibernate-orm/deployment`. E.g.:

```
./mvnw -T C1 -DskipDocs -DskipTests -DskipITs -DskipExtensionValidation -Dskip.gradle.tests clean install
./mvnw clean install -pl extensions/hibernate-orm/deployment,extensions/hibernate-orm/runtime -Dtest-containers -Dstart-containers
```

You will get failures like this:

```
[INFO] Running io.quarkus.hibernate.orm.singlepersistenceunit.SinglePersistenceUnitResourceInjectionEntityManagerTest
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.234 s <<< FAILURE! - in io.quarkus.hibernate.orm.singlepersistenceunit.SinglePersistenceUnitResourceInjectionEntityManagerTest
[ERROR] io.quarkus.hibernate.orm.singlepersistenceunit.SinglePersistenceUnitResourceInjectionEntityManagerTest  Time elapsed: 1.234 s  <<< ERROR!
java.lang.RuntimeException: io.quarkus.bootstrap.BootstrapException: Failed to create the application model for null
        at io.quarkus.test.QuarkusUnitTest.beforeAll(QuarkusUnitTest.java:544)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeBeforeAllCallbacks$8(ClassBasedTestDescriptor.java:368)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeBeforeAllCallbacks(ClassBasedTestDescriptor.java:368)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:192)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:78)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:136)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
        at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
        at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
        at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:188)
        at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:154)
        at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:128)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
        at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
        at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
Caused by: io.quarkus.bootstrap.BootstrapException: Failed to create the application model for null
        at io.quarkus.bootstrap.BootstrapAppModelFactory.resolveAppModel(BootstrapAppModelFactory.java:325)
        at io.quarkus.bootstrap.app.QuarkusBootstrap.bootstrap(QuarkusBootstrap.java:169)
        at io.quarkus.test.QuarkusUnitTest.beforeAll(QuarkusUnitTest.java:494)
        ... 40 more
Caused by: java.lang.IllegalStateException: Failed to parse command line arguments [clean, install, -pl, extensions/hibernate-orm/deployment,extensions/hibernate-orm/runtime, -Dtest-containers, -Dstart-containers]
        at io.quarkus.bootstrap.resolver.maven.options.BootstrapMavenOptions.invokeParser(BootstrapMavenOptions.java:233)
        at io.quarkus.bootstrap.resolver.maven.options.BootstrapMavenOptions.parse(BootstrapMavenOptions.java:98)
        at io.quarkus.bootstrap.resolver.maven.options.BootstrapMavenOptions.newInstance(BootstrapMavenOptions.java:119)
        at io.quarkus.bootstrap.resolver.maven.options.BootstrapMavenOptions.newInstance(BootstrapMavenOptions.java:115)
        at io.quarkus.bootstrap.resolver.maven.BootstrapMavenContextConfig.getInitializedCliOptions(BootstrapMavenContextConfig.java:236)
        at io.quarkus.bootstrap.resolver.maven.BootstrapMavenContextConfig.getPomForDirOrNull(BootstrapMavenContextConfig.java:184)
        at io.quarkus.bootstrap.BootstrapAppModelFactory.createBootstrapMavenContext(BootstrapAppModelFactory.java:217)
        at io.quarkus.bootstrap.BootstrapAppModelFactory.loadWorkspace(BootstrapAppModelFactory.java:337)
        at io.quarkus.bootstrap.BootstrapAppModelFactory.resolveAppModel(BootstrapAppModelFactory.java:262)
        ... 42 more
Caused by: java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at io.quarkus.bootstrap.resolver.maven.options.BootstrapMavenOptions.invokeParser(BootstrapMavenOptions.java:229)
        ... 50 more
Caused by: java.lang.OutOfMemoryError: Metaspace
```

Unfortunately, running a single test in isolation doesn't trigger the `OOME`; you need to run all tests in `hibernate-orm/deployment`.

**Comment by Sanne** [comment]
Ok, thanks @yrodiere . As mentioned on zulip, the leak is most likely related to the Narayana threads not being shut down - but that's obviously not an issue introduced in this PR.

**Comment by yrodiere** [comment]
Rebased on `main` now that #17279 was merged.

**Comment by stuartwdouglas** [comment]
https://github.com/quarkusio/quarkus/pull/17305 should allow the tests to pass

https://github.com/quarkusio/quarkus/issues/17304 is the underlying issue (well one of them), but the PR above works around it somewhat. Starting the PU's in parallel means that the work is moved off the main thread, so the ThreadLocal leak does not matter as much (as it only leaks to application threads, which are scoped to the life of the application).

**Comment by yrodiere** [comment]
That was quick! Thanks @stuartwdouglas . I'll rebase once #17305 gets merged.

**Comment by Sanne** [comment]
many thanks @stuartwdouglas ! Checking..

**Comment by yrodiere** [comment]
Rebased on master now that Stuart's CL leak fix was merged. Waiting for CI...

**Comment by Sanne** [comment]
Looks like we're still having the same problem :(

I'd be inclined to apply a workaround as it's not the fault of this PR - perhaps reconfiguring for more memory?

**Comment by yrodiere** [comment]
> I'd be inclined to apply a workaround as it's not the fault of this PR - perhaps reconfiguring for more memory?

Agreed. I tried raising the memory from 1500m to 2g. Let's see how it goes. I also added a comment to #17304 so that we remember to revert this at some point.

**Comment by famod** [comment]
A small warning here regarding that memory bump: the windows jvm job seems very sensitive to that. I needed a couple of iterations a few weeks ago until I got it somewhat stable...

**Comment by stuartwdouglas** [comment]
I will have another look then. It definitely worked on my machine with my PR, but maybe this was only a partial fix so it can still fail in some environments.

**Comment by stuartwdouglas** [comment]
I have opened https://github.com/quarkusio/quarkus/pull/17334 which has some more leak fixes, although some stuff still leaks because of the ThreadLocal issue with the Argoal Pool. I could work around this as well, but it would be a bit hacky, so we will see if the current fixes are enough.

**Comment by yrodiere** [comment]
Superseded by #17334

**Pull request closed without merging**
